### PR TITLE
Add fastq parser

### PR DESCRIPTION
**Pull request by Koeng101**
This is a fastq parser. It'll parse fastq files, such as the ones output by nanopore sequencers.

Fastq files have a bit more structure than fasta files, so I mostly just used the fasta file parsing method developed by @soypat (Thank you soypat for the #284 ! Was great and improved on my work!) with a few modifications to make it simpler (no more parse loop).

### CODE REVIEW

File: io/fastq/fastq.go [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+	if handleErr(err) != nil {
+		return Fastq{}, totalRead, handleErr(err)
+	}
+	line = line[:len(line)-1] // Exclude newline delimiter.
+	if string(line)[0] == '@' {
+		lookingForIdentifier = false
+	}
+	seqIdentifier = strings.Split(string(line), " ")[0][1:]
+	optionals = make(map[string]string)
+	for _, optionalDatum := range strings.Split(string(line), " ")[1:] {
```
Comment (matiasinsaurralde) [inline io/fastq/fastq.go:179]: We already call `strings.Split(string(line), " ")` in line 177 so I believe an initial optimization could be to do something like this and reuse the output from `strings.Split`:
```go
	splits := strings.Split(string(line), " ")
	seqIdentifier = splits[0][1:]
	optionals = make(map[string]string)
	for _, optionalDatum := range splits[1:] {
		optionalKey := strings.Split(optionalDatum, "=")[0]
		optionalValue := strings.Split(optionalDatum, "=")[1]
		optionals[optionalKey] = optionalValue
	}
```

Some benchmarks here -also results in less memory usage- 👍 
```
% go test -bench . -benchmem
goos: darwin
goarch: arm64
pkg: github.com/TimothyStiles/poly/io/fastq
BenchmarkParse-8   	  138547	      8812 ns/op	   74432 B/op	      93 allocs/op
PASS
ok  	github.com/TimothyStiles/poly/io/fastq	2.233s
% go test -bench . -benchmem
goos: darwin
goarch: arm64
pkg: github.com/TimothyStiles/poly/io/fastq
BenchmarkParse-8   	  145696	      7622 ns/op	   73088 B/op	      85 allocs/op
PASS
ok  	github.com/TimothyStiles/poly/io/fastq	2.457s
```

We can achieve the same in the `optionalDatum` block:
```go
	line = line[:len(line)-1] // Exclude newline delimiter.
	if string(line)[0] == '@' {
		lookingForIdentifier = false
	}
	lineSplits := strings.Split(string(line), " ")
	seqIdentifier = lineSplits[0][1:]
	optionals = make(map[string]string)
	for _, optionalDatum := range lineSplits[1:] {
		optionalSplits := strings.Split(optionalDatum, "=")
		optionalKey := optionalSplits[0]
		optionalValue := optionalSplits[1]
		optionals[optionalKey] = optionalValue
	}
```

And results in the following (reduces mem allocs by almost 50%, speed around 30%):
```
% go test -bench . -benchmem
goos: darwin
goarch: arm64
pkg: github.com/TimothyStiles/poly/io/fastq
BenchmarkParse-8   	  173823	      6780 ns/op	   72192 B/op	      57 allocs/op
PASS
ok  	github.com/TimothyStiles/poly/io/fastq	2.774s
```

**Comment by Koeng101** [comment]
Holy shit, that was fast - thank you! I've updated the code!

File: io/fastq/fastq.go [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+		}
+		fastqs = append(fastqs, fastq)
+	}
+	return fastqs, nil
+}
+
+// ParseByteLimited parses fastqs until byte limit is reached.
+// This is NOT a hard limit. To set a hard limit on bytes read use a
+// io.LimitReader to wrap the reader passed to the Parser.
+func (parser *Parser) ParseByteLimited(byteLimit int64) (fastqs []Fastq, bytesRead int64, err error) {
```
Comment (TimothyStiles) [inline io/fastq/fastq.go:105]: needs to be tested.

File: io/fastq/fastq.go [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+		fastq, _, err := parser.ParseNext()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				err = nil // EOF not treated as parsing error.
+			}
+			return fastqs, err
+		}
+		fastqs = append(fastqs, fastq)
+	}
+	return fastqs, nil
```
Comment (TimothyStiles) [inline io/fastq/fastq.go:99]: this line isn't touched by tests?

File: io/fastq/fastq.go [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+		Sequence:   *(*string)(unsafe.Pointer(&sequence)), // Stdlib strings.Builder.String() does this so it *should* be safe.
+	}
+	// Gotten to this point err is non-nil only in EOF case.
+	// We report this error to note the fastq may be incomplete/corrupt
+	// like in the case of using an io.LimitReader wrapping the underlying reader.
+	return fastq, totalRead, err
+}
+
+// Reset discards all data in buffer and resets state.
+func (parser *Parser) Reset(r io.Reader) {
```
Comment (TimothyStiles) [inline io/fastq/fastq.go:235]: should be tested.

File: io/fastq/fastq.go [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+}
+
+/******************************************************************************
+
+Start of  Read functions
+
+******************************************************************************/
+
+// ReadGz reads a gzipped  file into an array of Fastq structs.
+func ReadGz(path string) ([]Fastq, error) {
```
Comment (TimothyStiles) [inline io/fastq/fastq.go:247]: needs to be tested.

**Comment by Koeng101** [comment]
100% test coverage @Tim

**Pull request merged**